File: backend/services/workbench/validator.js
// ── Tool call argument validator ──
// Validates tool call arguments against the tool's JSON schema BEFORE execution.
// When invalid, returns a structured error message that can be fed back to the
// model as a tool result, triggering M2.7's self-correction loop.

function recordValidationFailure(toolName, args, error) {
    try {
        const { recordToolFailure } = require('../memory/tool-failure-memory');
        recordToolFailure({ toolName, args, error, phase: 'validation' });
    } catch (e) {
        // Failure learning is advisory and must never block validation.
    }
}

function validationFailure(toolName, args, error) {
    recordValidationFailure(toolName, args, error);
    return { valid: false, error };
}

function validateToolArguments(toolCall, toolDefinitions, messages = []) {
    const toolName = toolCall?.function?.name;
    if (!toolName) return { valid: false, error: 'Tool call is missing a function name.' };

    const tool = toolDefinitions?.find(t =>
        (t.function?.name || t.name) === toolName
    );
    if (!tool) return { valid: true }; // unknown tool — can't validate, pass through

    const schema = tool.function?.parameters || tool.input_schema;
    if (!schema) return { valid: true }; // no schema to validate against

    let args;
    try {
        args = JSON.parse(toolCall.function?.arguments || '{}');
    } catch (e) {
        return validationFailure(
            toolName,
            {},
            `Tool arguments for '${toolName}' are not valid JSON: ${e.message}. ` +
            `Arguments received: ${toolCall.function?.arguments}`
        );
    }

    // Compatibility shim for stale WebFetch/WebSearch tool schemas seen by some
    // third-party Claude clients. If the model sends `prompt` instead of the
    // proxy's canonical `url` / `query`, normalize it before schema checks.
    if (toolName === 'WebFetch' || toolName === 'web_fetch' || toolName === 'mcp__workspace__web_fetch') {
        if ((args.url === undefined || args.url === null || args.url === '') && typeof args.prompt === 'string' && args.prompt.trim()) {
            args.url = args.prompt.trim();
        }
    }
    if (toolName === 'WebSearch' || toolName === 'web_search' || toolName === 'mcp__workspace__web_search') {
        if ((args.query === undefined || args.query === null || args.query === '') && typeof args.prompt === 'string' && args.prompt.trim()) {
            args.query = args.prompt.trim();
        }
    }

    // ── The Proxy Execution Gate (Plan check) ──
    // Block code generation / file writes if a plan hasn't been formally established
    const mutatingTools = [
        'StrReplaceEditTool', 'BashTool',
        'mcp__filesystem__write_file', 'mcp__filesystem__create_directory', 
        'mcp__filesystem__move_file', 'mcp__filesystem__edit_file',
        'mcp__bash__execute_command', 'mcp__google_drive__write_file'
    ];
    if (mutatingTools.includes(toolName)) {
        // Check if plan.md exists anywhere in the conversation history
        const hasPlanned = messages.some(m => {
            if (typeof m.content === 'string') return m.content.toLowerCase().includes('plan.md');
            if (Array.isArray(m.content)) {
                return m.content.some(c => 
                    (typeof c.text === 'string' && c.text.toLowerCase().includes('plan.md')) ||
                    (c.type === 'tool_result' && typeof c.content === 'string' && c.content.toLowerCase().includes('plan.md')) ||
                    (c.type === 'tool_use' && JSON.stringify(c.input || {}).toLowerCase().includes('plan.md'))
                );
            }
            return false;
        });

        if (!hasPlanned) {
            return validationFailure(
                toolName,
                args,
                `PROXY GATE BLOCKED EXECUTION: You attempted to use a mutating tool ('${toolName}') without first establishing a plan.md file. You MUST read or write plan.md to solidify your architecture before executing code changes. This is a hard rule.`
            );
        }
    }

    // Check required fields from the schema
    const required = schema.required || [];
    const missing = required.filter(field => !(field in args));
    if (missing.length > 0) {
        const paramDescriptions = schema.properties
            ? missing.map(f => {
                const def = schema.properties[f];
                return `'${f}' (${def?.type || 'any'}${def?.description ? ': ' + def.description : ''})`;
            }).join(', ')
            : missing.join(', ');
        return validationFailure(
            toolName,
            args,
            `Missing required argument(s) for '${toolName}': ${paramDescriptions}. ` +
            `All required fields: ${required.join(', ')}.`
        );
    }

    // Check for unknown properties if schema has additionalProperties: false
    if (schema.additionalProperties === false && schema.properties) {
        const allowed = Object.keys(schema.properties);
        const unknown = Object.keys(args).filter(k => !allowed.includes(k));
        if (unknown.length > 0) {
            return validationFailure(
                toolName,
                args,
                `Unknown argument(s) for '${toolName}': ${unknown.join(', ')}. ` +
                `Allowed fields: ${allowed.join(', ')}.`
            );
        }
    }

    return { valid: true };
}

function buildValidationErrorToolMessage(toolCallId, toolName, errorMsg) {
    return {
        role: 'tool',
        tool_call_id: toolCallId,
        content: `[Validation Error] Tool call to '${toolName}' was rejected before execution:\n` +
                 `${errorMsg}\n\n` +
                 `[Proxy Self-Heal]: Fix the tool call arguments and retry. ` +
                 `Read the schema description above carefully. Do NOT stop.`
    };
}

module.exports = { validateToolArguments, buildValidationErrorToolMessage };

File: backend-py/app/services/workbench/validator.py
"""
Tool argument validation — validates tool calls against JSON schemas
before execution.

Port of backend/services/workbench/validator.js (136 lines).
"""

from __future__ import annotations

import json
import re


def validateToolArguments(
    toolCall: dict[str, object],
    toolDefinitions: list[dict[str, object]],
    messages: list[dict[str, object]] | None = None,
) -> dict[str, object]:
    """Validate a tool call against its schema.

    Args:
        tool_call: The tool call to validate. OpenAI format with
            ``function.name`` and ``function.arguments``.
        tool_definitions: List of known tool definitions.
        messages: Optional conversation history (used for plan-mode gating).

    Returns:
        ``{"valid": True}`` or ``{"valid": False, "error": "..."}``.
    """
    if messages is None:
        messages = []
    func = toolCall.get('function', {})
    if isinstance(func, dict):
        tool_name = func.get('name') or toolCall.get('name')
    else:
        tool_name = toolCall.get('name')
    if not isinstance(tool_name, str) or not tool_name:
        return {'valid': False, 'error': 'Missing tool name'}
    toolDef = _findToolDefinition(tool_name, toolDefinitions)
    if not toolDef:
        return {'valid': True}
    argsRaw = toolCall.get('input', '{}')
    if isinstance(func, dict):
        argsRaw = func.get('arguments', argsRaw)
    if isinstance(argsRaw, str):
        try:
            args = json.loads(argsRaw)
        except (json.JSONDecodeError, TypeError):
            return {'valid': False, 'error': f'Invalid JSON in arguments: {argsRaw[:200]}'}
    else:
        args = argsRaw
    if not isinstance(args, dict):
        return {'valid': False, 'error': f'Arguments must be a dict, got {type(args).__name__}'}
    schema = toolDef.get('parameters') or toolDef.get('input_schema')
    if isinstance(schema, dict):
        pass
    elif isinstance(toolDef.get('function'), dict):
        func = toolDef.get('function')
        if not isinstance(func, dict):
            schema = None
        else:
            schema = func.get('parameters')
    else:
        schema = None
    if not schema or not isinstance(schema, dict):
        return {'valid': True}
    args = _applyCompatibilityShims(tool_name, args)
    gateResult = _checkProxyExecutionGate(tool_name, args, messages)
    if not gateResult.get('valid'):
        return gateResult
    required = schema.get('required', [])
    if not isinstance(required, list):
        required = []
    for field in required:
        if not isinstance(field, str):
            continue
        if field not in args:
            return {'valid': False, 'error': f"Missing required field: '{field}'"}
        val = args[field]
        if val is None:
            return {'valid': False, 'error': f"Missing required field: '{field}'"}
        if isinstance(val, str) and (not val.strip()):
            return {'valid': False, 'error': f"Missing required field: '{field}'"}
    if schema.get('additionalProperties') is False:
        props = schema.get('properties', {})
        if not isinstance(props, dict):
            props = {}
        allowed = set(props.keys())
        extra = set(args.keys()) - allowed
        if extra:
            return {
                'valid': False,
                'error': f'Unknown fields: {", ".join(sorted(extra))}. Allowed fields: {", ".join(sorted(allowed))}',
            }
    properties = schema.get('properties', {})
    if not isinstance(properties, dict):
        properties = {}
    for field, value in args.items():
        propSchema = properties.get(field, {})
        if not isinstance(propSchema, dict):
            propSchema = {}
        propType = propSchema.get('type', '')
        if propType == 'string' and (not isinstance(value, str)):
            return {'valid': False, 'error': f"Field '{field}' must be a string"}
        if propType == 'integer' and (not isinstance(value, int)):
            return {'valid': False, 'error': f"Field '{field}' must be an integer"}
        if propType == 'number' and (not isinstance(value, (int, float))):
            return {'valid': False, 'error': f"Field '{field}' must be a number"}
        if propType == 'boolean' and (not isinstance(value, bool)):
            return {'valid': False, 'error': f"Field '{field}' must be a boolean"}
        if propType == 'array' and (not isinstance(value, list)):
            return {'valid': False, 'error': f"Field '{field}' must be an array"}
        if propType == 'object' and (not isinstance(value, dict)):
            return {'valid': False, 'error': f"Field '{field}' must be an object"}
    return {'valid': True}


def buildValidationErrorToolMessage(tool_call_id: str, tool_name: str, error_msg: str) -> dict[str, object]:
    """Build a tool result message for a validation error."""
    return {
        'tool_call_id': tool_call_id,
        'role': 'tool',
        'content': f"[Validation Error] Tool '{tool_name}' rejected before execution:\n{error_msg}\n\n[Proxy Self-Heal]: Fix the tool arguments and retry. Do NOT stop.",
    }


def _findToolDefinition(name: str, definitions: list[dict[str, object]]) -> dict[str, object] | None:
    """Find a tool definition by name, supporting both Anthropic and OpenAI formats."""
    for t in definitions:
        func = t.get('function', {})
        if isinstance(func, dict):
            tName = func.get('name') or t.get('name')
        else:
            tName = t.get('name')
        if tName == name:
            return t
    return None


def _applyCompatibilityShims(tool_name: str, args: dict[str, object]) -> dict[str, object]:
    """Apply compatibility shims for common tool name mappings."""
    if tool_name in ('WebFetch', 'web_fetch', 'mcp__workspace__web_fetch'):
        if 'prompt' in args and 'url' not in args:
            args = dict(args)
            args['url'] = args['prompt']
    if tool_name in ('WebSearch', 'web_search', 'mcp__workspace__web_search'):
        if 'prompt' in args and 'query' not in args:
            args = dict(args)
            args['query'] = args['prompt']
    return args


_MUTATINGToolPatterns = re.compile(
    '^(StrReplaceEditTool|BashTool|MCP.*(?:write|create|move|edit|delete|rename|copy)|mcp__.*(?:write|create|move|edit|delete))',
    re.IGNORECASE,
)


_PLAN_APPROVAL_PATTERNS = re.compile(
    r'(\.aug/plans/|plan\.md|plan_approved|Plan approved)',
    re.IGNORECASE,
)


def _checkProxyExecutionGate(
    tool_name: str, args: dict[str, object], messages: list[dict[str, object]]
) -> dict[str, object]:
    """Proxy Execution Gate: block mutating tools until a plan is approved.

    Checks for plan approval signals in the conversation context:
    - ``.aug/plans/`` path references (session plan files)
    - ``plan.md`` references (legacy plan files)
    - ``plan_approved`` or ``Plan approved`` markers
    """
    if not _MUTATINGToolPatterns.match(tool_name):
        return {'valid': True}
    for msg in messages:
        content = msg.get('content', '')
        if isinstance(content, str) and _PLAN_APPROVAL_PATTERNS.search(content):
            return {'valid': True}
        if isinstance(content, list):
            for block in content:
                if isinstance(block, dict):
                    text = block.get('text', '')
                    if isinstance(text, str) and _PLAN_APPROVAL_PATTERNS.search(text):
                        return {'valid': True}
    return {
        'valid': False,
        'error': f"Tool '{tool_name}' is blocked by the Proxy Execution Gate. No approved plan was found in the conversation. Create and approve a plan before using mutating tools.",
    }
